enhancements/substitutes-for.md

```diff
@@ -0,0 +1,379 @@
+plus `skips: A',A''`.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+The primary caveat with this design is how to handle the need for updating
+a single "replaces" reference to now reference multiple bundles.  For
+now this design proposes converting the additional replaces references
+into "skips", but there is an argument for instead moving the existing
+"replaces" reference to be a skips, and making the "replaces" reference
+point to the "newest" substitution.  That would at least ensure that
+if/when we must install a bundle as part of the path to upgrade to the
+latest, we install the most recently patched version of that bundle.
+
+However it adds additional complexity to determine which bundle is the
+"latest rebuild" of a given bundle, and more mutation to the CSV.
+
+I'd like to further discuss the implications of not doing this and what
+scenarios might be realistically problematic.
+
+### Risks and Mitigations
+
+The primary risk is that in mutating the CSVs we break the upgrade path in some
+way, either by not adding skips/replaces that we should, adding incorrect ones,
+or changing the upgrade behavior (such as the aforementioned cases where
+we go from a "replaces" upgrade path to a "skips" upgrade path, and some operator
+migration logic does not get executed).
+
+Not aware of any security concerns.
+
+
+## Design Details
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+Need to test the user scenarios described above.  Generally:
+
+1) rebuilding a bundle multiple times (creating transitive substitution
+relationships)
+2) rebuilding bundles that are referenced by other bundles
+
+And ensuring in both cases that the CSV content served by ListBundles
+contains all the necessary/correct skips/replaces references to ensure
+a correct upgrade graph for all channels.
+
+### Graduation Criteria
+
+This will be put directly into production once implemented, as it will be
+consumed by freshmaker.  So substitutes-for will need to be a GA api with
+a GA implementation in OPM.
+
+### Upgrade / Downgrade Strategy
+
+All logic related to this behavior will be tied to the opm binary used to
+generate the catalog, and the opm binary in the catalog image that is used
+to serve the content.  Therefore it should not require any changes to 
+OLM on the target cluster.  Catalogs produced that use this feature must
+work with any existing supported OCP cluster.
+
+Users who want to take advantage of this feature will need to use a current
+opm binary both to build their catalog DB as well as packaged in their 
+catalog image.
+
+Note:  Today the operator pipeline uses the latest opm binary to build catalog DBs
+(good) but matches the opm binary in the catalog to the ocp version.  We will
+need to either:
+
+1) Convince ourselves it is safe to always use a current opm binary in the catalog,
+even when the catalog is targeted at older clusters
```

Comment:
I don't see how this could work with respect to active changes to the registry (like adding new fields, for example). We are in a place where we have forced compatibility between 4.5-4.7, but that is only because of backports + lack of new features in 4.7. I think the only safe thing to do would be to backport this change to the older version of opm, otherwise we need to rethink how we ship opm in general as a separate entity

Reply:
I do think a larger discussion is in order about better defining what our DB + catalog api contract is(which i know has already been discussed a little elsewhere) so we can reason intelligently about whether we are breaking it or not (or more accurately, ensuring we do not break it w/o introducing meaningful new versions).  This doesn't have to be the forcing function for that discussion and if we think we can reasonably backport this as far as it needs to go w/o difficulty, i'm fine with that solution here.

Reply:
I guess if we're saying that "using a 4.7 binary for 4.5-4.7 catalogs" is safe today, i'd like to better understand what we envision doing in 4.8+ that will make it unsafe to use a 4.8+ binary on a 4.7 catalog/cluster, and whether we need to re-evaluate what those changes are/how they are made.

Reply:
It's no different than the problems we encountered when we first created the 4.6 version: it didn't work on 4.5 because we made schema changes. And that was reasonable, our new resolver needed different data than what was returned in 4.5 (basically the move to ListBundles to generate the cache). But it's the same problem in any future version: what happens when the cluster needs a new field that just doesn't get returned in an older version of the registry? We can definitely step through changes such that +2 versions still work, but we need some way to eventually bring in new required fields.

It's not a problem explicitly for 4.7->4.8, it's a generic problem for +2 future version of the registry whenever the cluster needs to have a new required thing. Right now, because our mandate has become "all versions of this thing must work with all clusters", we just have no way of introducing new requirements without backporting them all the way down.

Reply:
I guess it's a question of whether we ever "must" have that new field available to the cluster, or can the cluster fallback to some defaults if the fields are not provided.  

In any case it's more likely the opposite, right?  The catalog opm + DB would be ahead of the cluster version, so it would be returning fields the cluster didn't expect.  Again, whether that's a problem, or still results in safe behavior on an older cluster, is a bit difficult to reason about absent a specific example.  Certainly possible to envision both cases.